### ISSUE TITLE
Placeholders notes and known issues

### ISSUE THREAD

**Issue by colagrosso** [standardebooks/web#430]
The `placeholders` branch is in good shape and is ready for you to take over, Alex. I've tested it a few different ways, and I'm happy with it.

Three known issues:

### 1\. `Ebooks` SQL schema

In your existing database, you'll need to modify these columns in `Ebooks` table to allow nulls:

ALTER TABLE Ebooks MODIFY COLUMN \`WwwFilesystemPath\` varchar(511) NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`RepoFilesystemPath\` varchar(511) NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`Description\` text NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`LongDescription\` text NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`Language\` varchar(10) NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`WordCount\` int(10) unsigned NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`ReadingEase\` float NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`EbookCreated\` datetime NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`EbookUpdated\` datetime NULL;  
ALTER TABLE Ebooks MODIFY COLUMN \`TextSinglePageByteCount\` bigint unsigned NULL;

### 2\. PHPStan errors to fix later

Because those columns above can now be null, PHPStan is throwing errors like:

```
 ------ ----------------------------------------------------------------------------   
  Line   lib/AtomFeed.php                                                                                                                                       
 ------ ----------------------------------------------------------------------------   
  58     Cannot call method format() on Safe\\DateTimeImmutable|null.    
 ------ ----------------------------------------------------------------------------   

 ------ ----------------------------------------------------------------------------   
  Line   scripts/generate-bulk-downloads                                      
 ------ ----------------------------------------------------------------------------   
  113    Cannot call method format() on Safe\\DateTimeImmutable|null.          
  114    Cannot call method getTimestamp() on Safe\\DateTimeImmutable|null.    
 ------ ----------------------------------------------------------------------------   

 ------ ----------------------------------------------------------------------------   
  Line   templates/EbookGrid.php                                                        
 ------ ----------------------------------------------------------------------------   
  49     Parameter \#1 $num of function number\_format expects float, int|null given.    
 ------ ----------------------------------------------------------------------------   

 ------ ----------------------------------------------------------------------------   
  Line   templates/OpdsAcquisitionEntry.php                             
 ------ ----------------------------------------------------------------------------   
  27     Cannot call method format() on Safe\\DateTimeImmutable|null.    
  28     Cannot call method format() on Safe\\DateTimeImmutable|null.    
  29     Cannot call method format() on Safe\\DateTimeImmutable|null.    
 ------ ----------------------------------------------------------------------------   

 ------ ----------------------------------------------------------------------------   
  Line   templates/RssEntry.php                                         
 ------ ----------------------------------------------------------------------------   
  13     Cannot call method format() on Safe\\DateTimeImmutable|null.    
 ------ ----------------------------------------------------------------------------   

 ------ ----------------------------------------------------------------------------   
  Line   www/ebooks/get.php                                                            
 ------ ----------------------------------------------------------------------------   
  126    Parameter \#1 $num of function number\_format expects float, int|null given.    
  178    Cannot call method format() on Safe\\DateTimeImmutable|null.                   
  179    Cannot call method format() on Safe\\DateTimeImmutable|null.                   
 ------ ----------------------------------------------------------------------------   
```

These errors are on `EbookCreated`, `EbookUpdated`, and `WordCount`, which can now be null. I'll fix the errors eventually with more checks, but at the moment it didn't seem worth it because we knew these fields are OK for published ebooks.

### 3\. Sorting placeholders at the end in search results

Of the 4 sort orders on the search results page:

1. S.E. release date  
2. Author name  
3. Reading ease  
4. Length

Placeholders don't have any of those properties set. (They do have an author, but not an author's `SortName`.) Therefore I put the placeholders at the end of the search results. I had to modify the queries that used Author name and Length as sort orders to put them at the end. I think this is the right thing to do, I just wanted to call it out here since it was mixed in with other frontend changes.

### Getting started

To get started, you can create your own placeholders with the form at `/ebook-placeholders/new` or bulk add placeholders from the spreadsheet.

### Bulk adding placeholders from the spreadsheet

Use this gist for help: [https://gist.github.com/colagrosso/7fc986a0844157af7cd048f5f0018541](https://gist.github.com/colagrosso/7fc986a0844157af7cd048f5f0018541). It has 3 files:

1. [convert-ebook-placeholders-from-csv-to-sql](https://gist.github.com/colagrosso/7fc986a0844157af7cd048f5f0018541#file-convert-ebook-placeholders-from-csv-to-sql)   
2. [se-placeholders.csv](https://gist.github.com/colagrosso/7fc986a0844157af7cd048f5f0018541#file-se-placeholders-csv)   
3. [se-placeholders.sql](https://gist.github.com/colagrosso/7fc986a0844157af7cd048f5f0018541#file-se-placeholders-sql) 

You can download a CSV from the [spreadsheet](https://docs.google.com/spreadsheets/d/1OXHJcego9CINwWSJfEAo8aTJ7jZDbhvXshYsEcMuato/edit?usp=sharing). Note that the spreadsheet has two tabs ("List" and "Copyrighted works"). I combined the two tabs into a single CSV, `se-placeholders.csv`. You can use that or a subset of the CSV.

The script converts the (hardcoded) `se-placeholders.csv` file into a sequence of SQL statements to insert the placeholders into the DB. Run it like this:

``` 
./convert-ebook-placeholders-from-csv-to-sql > se-placeholders.sql  
mysql se < se-placeholders.sql  
``` 

The `se-placeholders.sql` in the gist is the full set of SQL statements for the combined CSV. You can spot check it to ensure that the statements look correct.

**Comment by acabal** [comment]
After a quick test:

- Missing SQL to add `CanCreateEbookPlaceholder` to `Benefits` table
- The `CanCreateEbookPlaceholder` column should be named `CanCreateEbookPlaceholders` for consistency 
- I added one placeholder for Agatha Christie, Poirot series. All other info was blank. But after it was added, the placeholder listed two authors, `A. Merritt` and `Agatha Christie`.

I stopped there for now.

**Comment by colagrosso** [comment]
Sorry for the trouble. 

You probably ran the SQL yourself but for future reference, here is the SQL I forgot to include here:

```sql
ALTER TABLE Benefits ADD COLUMN `CanCreateEbookPlaceholders` tinyint(1) unsigned NOT NULL DEFAULT 0 after `CanEditUsers`;
```

I just made `CanCreateEbookPlaceholders` plural in e3f0342e5587c7836b40e146d3fb1cd951ebfce0 and pushed that commit to the `placeholders` branch. If you pull from `placeholders`, you'll have to update your DB again.

That's a bummer that your first placeholder didn't work. I can't reproduce the problem yet.

It's a long shot, but is it possible that `A. Merritt` was a stale entry in the `Contributors` table from a previous ebook? Sometimes for testing I'll drop and recreate my `Ebooks` table but forget to do the same with `Contributors`, `CollectionEbooks`, etc.

Here's what I got from the Firefox and Chrome web developer tools Network menu when I submit the form:


Firefox:
```
-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="author-name-1"

Agatha Christie
-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="author-name-2"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="author-name-3"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="translator-name-1"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="translator-name-2"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="ebook-title"

A test placeholder
-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="ebook-placeholder-year-published"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="collection-name-1"

Poirot
-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="sequence-number-collection-name-1"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="collection-name-2"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="sequence-number-collection-name-2"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="collection-name-3"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="sequence-number-collection-name-3"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="ebook-placeholder-difficulty"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="ebook-placeholder-status"

wanted
-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="ebook-placeholder-transcription-url"


-----------------------------182822189336008064892524579620
Content-Disposition: form-data; name="ebook-placeholder-notes"


-----------------------------182822189336008064892524579620--
```

Chrome:
```
author-name-1: Agatha Christie
author-name-2: 
author-name-3: 
translator-name-1: 
translator-name-2: 
ebook-title: A test placeholder 2
ebook-placeholder-year-published: 
collection-name-1: Poirot
sequence-number-collection-name-1: 
collection-name-2: 
sequence-number-collection-name-2: 
collection-name-3: 
sequence-number-collection-name-3: 
ebook-placeholder-difficulty: 
ebook-placeholder-status: wanted
ebook-placeholder-transcription-url: 
ebook-placeholder-notes: 
```

So there was no extra author submitted by accident. I'll keep looking for a problem.

**Comment by acabal** [comment]
Might have been a stale entry but even if it was, why were two contributors added?

**Comment by acabal** [comment]
Yup, just tried again, added an ebook with Agatha Christie as the only author, but the created placeholder has two authors: A. N. Afanasyev and Agatha Christie

**Comment by colagrosso** [comment]
A stale entry in the `Contributors` table would mean that there would be two rows that match this query:

```sql
select * from Contributors where EbookId = 4000;
+---------+-----------------+-----------------+----------+--------------+----------+----------+-----------+-----------+
| EbookId | Name            | UrlName         | SortName | WikipediaUrl | MarcRole | FullName | NacoafUrl | SortOrder |
+---------+-----------------+-----------------+----------+--------------+----------+----------+-----------+-----------+
|    4000 | Agatha Christie | agatha-christie | NULL     | NULL         | aut      | NULL     | NULL      |         0 |
+---------+-----------------+-----------------+----------+--------------+----------+----------+-----------+-----------+
```

Could you take a look at your `Contributors` table? It's a long shot, but I'm wondering if there stale entries there. I don't think the placeholder form is adding the A. Merritt or A. N. Afanasyev rows. I'm guessing they were already there. If you run a

```
select * from Contributors;
```

before and after submitting the placeholder form, maybe you'll see what was added and what was already there.

**Comment by acabal** [comment]
OK, that seems to have been the problem. After truncating the database and starting from scratch it works. I'll continue to review.

**Comment by colagrosso** [comment]
Right on, thanks.

**Comment by acabal** [comment]
OK, I've pushed a commit that does some basic front-end formatting. I think this will be good for now.

Another thing I noticed is that placeholders appear in `/ebook` when no search parameters are set. I think we should not show placeholders in that case - only if the user has actually entered some parameters in the search form.

**Comment by colagrosso** [comment]
> Another thing I noticed is that placeholders appear in `/ebooks` when no search parameters are set. I think we should not show placeholders in that case - only if the user has actually entered some parameters in the search form.

Yep, agreed. I sent #431 to handle that.

New problem: Rendering on Chrome
![Screenshot_2024-12-11_23-57-17](https://github.com/user-attachments/assets/fa9c40aa-c2c5-4bd0-b837-712335768656)

You corrected this line in `templates/EbookGrid.php`:

```
<div class="placeholder-cover"></div>
```

to the self-closing tag:

```
<div class="placeholder-cover"/>
```

but that seems to be what's causing the problem on Chrome. Changing it back to `<div class="placeholder-cover"></div>` makes it work on Chrome.  

I know Firefox is the preferred browser, and the self-closing tag is valid XHTML. I'm out of my depth, so consider this just a friendly bug report, no stake in how to fix it. I can provide more details or try other fixes if it would help.

**Comment by acabal** [comment]
Thanks, I'll just change it back.

**Comment by acabal** [comment]
OK, everything looks good. I've merged this into master and it's now live on the site.

Excellent work!

I squashed the branch, so for further work please create a new branch off master.

Next up in order of priority:

- Authors who have only placeholder works end up at a 404. For example [The Maltese Falcon](https://standardebooks.org/ebooks/dashiell-hammet/the-maltese-falcon) works but [Dashiell Hammet](https://standardebooks.org/ebooks/dashiell-hammet) 404s. We should return a page listing all the placeholders.
- Quick fix: In the `EbookPlaceholderForm.php` template, there are several `<fieldset>`s  that only have a single child `<label>`. We should restructure that to remove those `<fieldset>`s as that doesn't make semantic sense. For example line 79, line 105, line 130 etc.
- When we release an ebook that was a placeholder, does the system delete the placeholder? It should.
- Update the wanted list to use placeholders from DB
- Form to update placeholders

**Comment by colagrosso** [comment]
Woohoo, congrats on the launch.

> When we release an ebook that was a placeholder, does the system delete the placeholder? It should.

Yes, it does.

I'll get on the other fixes you listed above. Two others:

- [ ] Mobile placeholder oulines are too big
- [ ] That [The Maltese Falcon](https://standardebooks.org/ebooks/dashiell-hammet/the-maltese-falcon) placeholder is in the wrong collection. It should be a member of [The Guardian’s Best 100 Novels in English (2015)](https://standardebooks.org/collections/the-guardians-best-100-novels-in-english-2015), but it's in a [collection](https://standardebooks.org/collections/guardians-best-100-novels-in-english-2015) lacking the initial "The." Sorry about that. I thought I fixed all of those in the spreadsheet, but I missed that one. I did a quick pass over the [collections](https://standardebooks.org/collections) page and didn't see any other problems.

To fix the Maltese Falcon collection data, you can run this SQL:

```sql
DELETE FROM CollectionEbooks WHERE EbookId = (SELECT EbookId FROM Ebooks WHERE Identifier = "url:https://standardebooks.org/ebooks/dashiell-hammet/the-maltese-falcon");

DELETE FROM Collections WHERE UrlName = 'guardians-best-100-novels-in-english-2015';

INSERT INTO CollectionEbooks (EbookId, CollectionId, SequenceNumber, SortOrder) VALUES ((SELECT EbookId FROM Ebooks WHERE Identifier = "url:https://standardebooks.org/ebooks/dashiell-hammet/the-maltese-falcon"), (SELECT CollectionId FROM Collections WHERE UrlName = "the-guardians-best-100-novels-in-english-2015"), 54, 0);
```

**Comment by acabal** [comment]
Thanks, I fixed it!!

**Comment by acabal** [comment]
Also, this just came up - when adding a new placeholder that creates a new collection, we need to have a dropdown to specify whether the collection is a `set` or a `series`. Currently `Collection.Type` is nullable but it should not be, I updated that in master. 

We can do something similar to what we did for death year of artists of an artwork - prefill from the collection name from the datalist; if the collection already exists then leave the type empty to not change it. If the collection does not exist then selecting a type, either `set` or `series`, is required.
